### PR TITLE
Implement `Branch.cherry_pick()`

### LINKED ISSUE #7346: [Hi-lvl Python SDK] Add `Branch.cherry_pick()`
It would be cool if I could cherry-pick commits from a branch in the lakeFS Python wrapper.

Maybe like this:

```py
import lakefs

b = lakefs.Branch("repo", "my-branch")
# I have some data on a branch named "previous" at HEAD.
b.cherry_pick("previous")
```

If this is feasible, its signature would probably be determined by the API. Input ref should be a `ReferenceType`, i.e., strs for ref expressions, proper `lakefs.Reference`s, and `Commit`s.

Also, on success, it should return the new commit on b (that is actually returned by the API as far as I can see, https://pydocs-sdk.lakefs.io/docs/BranchesApi.html#cherry_pick).

### PR DESCRIPTION
**Pull request by nicholasjng**
As a minimal wrapper around `BranchesApi.cherry_pick()`. Since the API returns the created commit , we return the API result directly without any postprocessing.

Closes #7346.

-------------------

For reviewers: I welcome guidance especially on the exception taxonomy - I just rolled with what similar APIs had in their docstrings (and the obvious cases from the SDK client doc).

### CODE REVIEW

**Comment by nicholasjng** [comment]
cc @itaiad200

**Review by N-o-Z — CHANGES_REQUESTED** [review]
Thanks @nicholasjng 

Looks great! Can you please just add integration tests for this operation, and test parent_number = None explicitly.
See [documentation](https://github.com/treeverse/lakeFS/blob/master/clients/python-wrapper/tests/integration/README.md) regarding running tests

File: clients/python-wrapper/lakefs/branch.py [hunk 1]
```diff
@@ -124,6 +124,24 @@ def get_commit(self):
...
+            The destination branch is parent 1, which is the default behaviour.
+        :return: The cherry-picked commit at the head of the branch.
+        :raise NotFoundException: If either the repository or target reference do not exist.
+        :raise NotAuthorizedException: If the user is not authorized to perform this operation.
+        :raise ServerException: For any other errors.
+        """
+        ref = reference if isinstance(reference, str) else reference.id
+        cherry_pick_creation = lakefs_sdk.CherryPickCreation(ref=ref, parent_number=parent_number)
+        with api_exception_handler():
+            return self._client.sdk_client.branches_api.cherry_pick(self._repo_id, self._id, cherry_pick_creation)
```
Comment (N-o-Z) [inline clients/python-wrapper/lakefs/branch.py:143]: You need to wrap this with the lakeFS wrapper, see `uncommitted` example

Comment (nicholasjng) [reply clients/python-wrapper/lakefs/branch.py]: Done in [01d9ada](https://github.com/treeverse/lakeFS/commit/01d9adab5d594458fe1ef742c19f109a5233b3ea).

**Comment by nicholasjng** [comment]
I added a first integration test in [01d9ada](https://github.com/treeverse/lakeFS/commit/01d9adab5d594458fe1ef742c19f109a5233b3ea). What do you want to be the outcome of the parent_number = None test?

**Comment by N-o-Z** [comment]
Thanks @nicholasjng,

Testing parent == None protects us against any breaking changes in the underlying SDK

**Review by N-o-Z — APPROVED** [review]
LGTM!

**Comment by N-o-Z** [comment]
@nicholasjng Your test seem to fail. Would you like to debug it or do I have your permission to push fixes to the branch?

**Comment by nicholasjng** [comment]
Sure, not sure how this is related to the changes though - seems like it could not find a Docker image?

**Comment by N-o-Z** [comment]
> Sure, not sure how this is related to the changes though - seems like it could not find a Docker image?

The failure you see is actually unrelated to my comment and is probably transient.
Unfortunately, we can't run our pipeline tests on forks due to security concerns. I've ran the tests locally and  the cherry-pick test failed.

File: clients/python-wrapper/lakefs/branch.py [hunk 2]
```diff
@@ -124,6 +124,25 @@ def get_commit(self):
...
         return super().get_commit()
 
+    def cherry_pick(self, reference: ReferenceType, parent_number: Optional[int] = None) -> Commit:
+        """
+        Cherry-pick a given reference onto the branch.
+
+        :param reference: ID of the reference to cherry-pick.
+        :param parent_number: When cherry-picking a merge commit,
+            the parent number (starting from 1) relative to which to perform the operation.
+            The destination branch is parent 1, which is the default behaviour.
```
Comment (arielshaqed) [inline clients/python-wrapper/lakefs/branch.py:134]: Not sure what this means.  Either I need more coffee or
```suggestion
            The default branch is parent 1, which is the default behaviour.
```
(or both).

File: clients/python-wrapper/lakefs/branch.py [hunk 3]
```diff
@@ -124,6 +124,25 @@ def get_commit(self):
...
+        :return: The cherry-picked commit at the head of the branch.
+        :raise NotFoundException: If either the repository or target reference do not exist.
+        :raise NotAuthorizedException: If the user is not authorized to perform this operation.
+        :raise ServerException: For any other errors.
+        """
+        ref = reference if isinstance(reference, str) else reference.id
```
Comment (arielshaqed) [inline clients/python-wrapper/lakefs/branch.py:140]: @N-o-Z this line be a candidate for refactoring

**Review by arielshaqed — COMMENTED** [review]
**Nice!**  Minor quibble about docstring.

File: clients/python-wrapper/lakefs/branch.py (hunk 2 above)
Comment (nicholasjng) [reply clients/python-wrapper/lakefs/branch.py:134]: This comes straight from the docstring of `lakefs_sdk.CherryPickCreation` fwiw.

**Comment by nicholasjng** [comment]
Seems like the culprit was cherry-pick not understanding a raw branch name as argument.

I fixed the test by supplying the HEAD reference instead, and tweaked some of the asserts after understanding the workings of cherry-picking better.

File: clients/python-wrapper/lakefs/branch.py (hunk 2 above)
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/branch.py:134]: @nicholasjng I tend to agree with @arielshaqed about this one and give you permission to modify the description in our swagger.yaml
I suggest the following:

> When cherry-picking a merge commit, the parent number (starting from 1) with which to perform the diff.
The default branch is parent 1.

File: clients/python-wrapper/lakefs/branch.py [hunk 4]
```diff
@@ -124,6 +124,25 @@ def get_commit(self):
         self._commit = None
         return super().get_commit()
 
+    def cherry_pick(self, reference: ReferenceType, parent_number: Optional[int] = None) -> Commit:
```
Comment (N-o-Z) [inline clients/python-wrapper/lakefs/branch.py:127]: Can we add the force flag as well (with default to False)

**Review by N-o-Z — CHANGES_REQUESTED** [review]
@nicholasjng sorry to be dragging this a bit more, can you take another look at the comments

File: clients/python-wrapper/lakefs/branch.py (hunk 4 above)
Comment (nicholasjng) [reply clients/python-wrapper/lakefs/branch.py:127]: Sure. What does it actually do that a normal cherry-pick does not (there is no doc in the SDK argument)?

Comment (N-o-Z) [reply clients/python-wrapper/lakefs/branch.py]: @nicholasjng on second thought - please ignore this comment. This flag is related to a feature which is currently in development. We'll add it in the future if needed.

**Comment by nicholasjng** [comment]
> @nicholasjng sorry to be dragging this a bit more, can you take another look at the comments

Done, please take another look.

File: clients/python-wrapper/lakefs/branch.py (hunk 4 above)
Comment (nicholasjng) [reply clients/python-wrapper/lakefs/branch.py]: So remove the `force` argument again?

File: clients/python-wrapper/lakefs/branch.py (hunk 2 above)
Comment (arielshaqed) [reply clients/python-wrapper/lakefs/branch.py:134]: # 😳

**Comment by nicholasjng** [comment]
As a compromise, I checked in the `force` option in [1b286e2](https://github.com/treeverse/lakeFS/pull/7348/commits/1b286e2c5463b8ccc82b67a82bfde2c9017070e9), and reverted it in [f79ed3f](https://github.com/treeverse/lakeFS/pull/7348/commits/f79ed3ff2ccff5facc882136336b5eb030d8bd1e), so that you only need to roll forward [f79ed3f](https://github.com/treeverse/lakeFS/pull/7348/commits/f79ed3ff2ccff5facc882136336b5eb030d8bd1e) once the force option goes live.

**Comment by nicholasjng** [comment]
> @nicholasjng You updated the swagger but forgot to update the docstring in the method itself :)

Fixed.
